src/main/cpp/tests/cast_string.cpp

```diff
@@ -473,3 +481,138 @@ TEST_F(StringToDecimalTests, Edges)
+                                        "0001",
+                                        "0000.123",
+                                        "123",
+                                        "123.45",
+                                        "45.123",
+                                        "-45.123",
+                                        "0.45123",
+                                        "-0.45123",
+                                        "999999999999999999999",
+                                        "99999999999999999999",
+                                        "9999999999999999999",
+                                        "18446744073709551609",
+                                        "18446744073709551610",
+                                        "18446744073709551619999999999999",
+                                        "-18446744073709551609",
+                                        "-18446744073709551610",
+                                        "-184467440737095516199999999999997"};
+
+  std::vector<bool> valids{1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1};
+
+  auto expected =
+    cudf::strings::to_floats(strings_column_view(in), cudf::data_type{type_to_id<TypeParam>()});
+  expected->set_null_mask(cudf::test::detail::make_null_mask(valids.begin(), valids.end()));
+
+  auto const result = spark_rapids_jni::string_to_float(
+    data_type{type_to_id<TypeParam>()}, strings_column_view{in}, false, rmm::cuda_stream_default);
+
+  CUDF_TEST_EXPECT_COLUMNS_EQUAL(result->view(), expected->view());
+}
+
+TYPED_TEST(StringToFloatTests, InfNaN)
+{
+  cudf::test::strings_column_wrapper in{"NaN", "-Infinity", "inf", "Infinity", "-inf", "-nan"};
+
+  std::vector<bool> valids{1, 1, 1, 1, 1, 0};
+
+  auto expected =
+    cudf::strings::to_floats(strings_column_view(in), cudf::data_type{type_to_id<TypeParam>()});
+  expected->set_null_mask(cudf::test::detail::make_null_mask(valids.begin(), valids.end()));
+
+  auto const result = spark_rapids_jni::string_to_float(
+    data_type{type_to_id<TypeParam>()}, strings_column_view{in}, false, rmm::cuda_stream_default);
+
+  CUDF_TEST_EXPECT_COLUMNS_EQUAL(result->view(), expected->view());
+}
+
+TYPED_TEST(StringToFloatTests, InvalidValues)
+{
+  cudf::test::strings_column_wrapper in{"A", "null", "na7.62", "e", ".", "", "f"};
+
+  std::vector<bool> valids{0, 0, 0, 0, 0, 0, 0, 0};
+
+  auto expected =
+    cudf::strings::to_floats(strings_column_view(in), cudf::data_type{type_to_id<TypeParam>()});
+  expected->set_null_mask(cudf::test::detail::make_null_mask(valids.begin(), valids.end()));
+
+  auto const result = spark_rapids_jni::string_to_float(
+    data_type{type_to_id<TypeParam>()}, strings_column_view{in}, false, rmm::cuda_stream_default);
+
+  CUDF_TEST_EXPECT_COLUMNS_EQUAL(result->view(), expected->view());
+}
+
+TYPED_TEST(StringToFloatTests, ANSIInvalids)
+{
+  cudf::test::strings_column_wrapper in[] = {{"A"}, {"."}, {"e"}};
```

Comment:
When are we going to support hex as input? Not exactly the same here, but it is something that Spark supports and we cannot/do not support.

ie `0x1p0 => 2.0`

Reply:
We can always construct a string from hex, but with some simple hacks (the working code may not be exactly like this):
```
char data[][] { { 0x0a, 0x0b, 0x01, 0x04 }, { 0x0a, 0x0b, 0x01, 0x04 }, .....};
cudf::test::strings_column_wrapper in[] = { {std::string(data[0], 4)}, {std::string(data[1], 4)}, .....};
```

Reply:
If you want to use hex frequently, you can write a utility function for simplifying the work in test too.

Reply:
The reason I ask is that I thought one of the goals of this was to fix the issues with casting a string to floating point.  It is off by default because of these issues.  Many of them are corner cases that CUDF does not support, but we need to in order to have this operation on by default.

Hex strings is one of the reasons it is off by default. I didn't see any tests for hex strings so that is why I asked. I probably put this comment in the wrong place with the ASCII test. It is not really related.

```
        willNotWorkOnGpu("Currently hex values aren't supported on the GPU. Also note " +
            "that casting from string to float types on the GPU returns incorrect results when " +
            "the string represents any number \"1.7976931348623158E308\" <= x < " +
            "\"1.7976931348623159E308\" and \"-1.7976931348623159E308\" < x <= " +
            "\"-1.7976931348623158E308\" in both these cases the GPU returns Double.MaxValue " +
            "while CPU returns \"+Infinity\" and \"-Infinity\" respectively. To enable this " +
            s"operation on the GPU, set ${RapidsConf.ENABLE_CAST_STRING_TO_FLOAT} to true.")
```

If we are not fixing these issues can we have a follow on issue filed to fix them, now that we have a custom kernel? Also what are the performance differences vs the existing kernel. The bug just listed this as a performance fix, so if that is the case then have we measured the performance improvement vs the existing code? I assume it is a lot better because it is a single kernel instead of multiple etc..., but it would be nice to know.

If we did fix the issues can we have some tests to verify that we did?

Reply:
Sorry I misunderstood what you asked. I meant we can **test** string constructed from hex **value** in the code (like the snippet above: `string` from `char` given as hex values). For converting hex string like `0x41973333` then we didn't support yet and definitely need a custom kernel.

Reply:
@revans2 I just realized that cudf already has `hex_to_integers` API that can convert strings into integer numbers. I believe that we can just do a `bit_cast` from the output integers into the target doubles, which is very straightforward. If there is anything missing here then we can easily add that.

Reply:
Not really. The hex format is not just parse the hex to get the bit representation.  The values fall under a specific format that appears to be the same for both C, C++, and Java

https://www.exploringbinary.com/hexadecimal-floating-point-constants/  

From what I can tell the form is `0x[HEX VALUE]p[EXP]` where the result is the parsed `HEX VALUE * 2 ^ EXP`. and oddly the `EXP` is represented in decimal not hex.

Also we have the issue that we need to be able to support these intermixed in the same column.

`["1.0", "0x1p0"]` so we would need to move that kind of logic into our custom kernel.

Reply:
Added an issue for this.